crates/bevy_ecs/src/system/const_param_checking.rs

```diff
@@ -0,0 +1,526 @@
+                        access,
+                        name,
+                    } => {
+                        if *type_id_self != *type_id {
+                            return None;
+                        }
+                        if matches!(access_self, AccessType::Mut) {
+                            if let Some(name_self) = name_self {
+                                let name_self = *name_self;
+                                if let Some(name) = name {
+                                    let name = *name;
+                                    return Some(SystemPanicMessage {
+                                        lhs_access_type: *access_self,
+                                        lhs_name: name_self,
+                                        rhs_access_type: *access,
+                                        rhs_name: name,
+                                    });
+                                }
+                                return Some(SystemPanicMessage {
+                                    lhs_access_type: *access_self,
+                                    lhs_name: "",
+                                    rhs_access_type: *access,
+                                    rhs_name: "",
+                                });
+                            }
+                        } else {
+                            if matches!(access, AccessType::Mut) {
+                                if let Some(name_self) = name_self {
+                                    let name_self = *name_self;
+                                    if let Some(name) = name {
+                                        let name = *name;
+                                        return Some(SystemPanicMessage {
+                                            lhs_access_type: *access_self,
+                                            lhs_name: name_self,
+                                            rhs_access_type: *access,
+                                            rhs_name: name,
+                                        });
+                                    }
+                                    return Some(SystemPanicMessage {
+                                        lhs_access_type: *access_self,
+                                        lhs_name: "",
+                                        rhs_access_type: *access,
+                                        rhs_name: "",
+                                    });
+                                }
+                            }
+                        }
+                        None
+                    }
+                }
+            }
+        }
+    }
+}
+/// The type of access to a component - either read-only reference or mutable reference.
+#[derive(Copy, Clone, Debug)]
+pub enum AccessType {
+    /// Immutable reference access (&T)
+    Ref,
+    /// Mutable reference access (&mut T)
+    Mut,
+}
+
+/// A tree structure representing the `With<T>`, `Added<T>`, and `Changed<T>` filters
```

Comment:
If I understand correctly, these are semantically unordered sets, and are implemented as trees rather than lists to allow efficient concatenation.  Is that right?  It might be useful to have a comment to that effect.